packages/block-editor/src/components/list-view/use-list-view-open-selected-item.js

```diff
@@ -0,0 +1,139 @@
+			: null;
+
+	// Track the expanded state of any parents.
+	// To calculate the number of expanded items correctly.
+	let parentExpandedState = null;
+	if ( parentClientIds ) {
+		parentExpandedState = expandedState[ parentClientIds[ 0 ] ];
+	}
+
+	useEffect( () => {
+		// If the selectedTreeId is the same as the selected block,
+		// it means that the block was selected using the block list tree.
+		if ( selectedTreeId === firstSelectedBlockClientId ) {
+			return;
+		}
+
+		// If the selected block has parents, get the top-level parent.
+		if ( parentClientIds ) {
+			// If the selected block has parents,
+			// expand the tree branch.
+			setExpandedState( {
+				type: 'expand',
+				clientIds: selectedBlockParentClientIds,
+			} );
+		}
+	}, [ firstSelectedBlockClientId ] );
+
+	useLayoutEffect( () => {
+		// If the selectedTreeId is the same as the selected block,
+		// it means that the block was selected using the block list tree.
+		if ( selectedTreeId === firstSelectedBlockClientId ) {
+			return;
+		}
+		if (
+			scrollContainer &&
+			!! firstSelectedBlockClientId &&
+			Array.isArray( clientIdsTree ) &&
+			clientIdsTree.length
+		) {
+			// Grab the selected id. This is the point at which we can
+			// stop counting blocks in the tree.
+			let selectedId = firstSelectedBlockClientId;
+
+			// If the selected block has parents, get the top-level parent.
+			if ( parentClientIds ) {
+				selectedId = parentClientIds[ 0 ];
+				// If the selected block has parents,
+				// check to see if the selected tree is expanded
+				// so we can accurately calculate the scroll container top value.
+				if ( ! parentExpandedState ) {
+					return;
+				}
+			}
+
+			// Count expanded blocks in the tree up until the selected block,
+			// so we can calculate the scroll container top value.
+			let listItemHeightFactor = 0;
+			clientIdsTree.every( ( item ) => {
+				if ( item?.clientId === selectedId ) {
+					return false;
+				}
+				listItemHeightFactor += countBlocks( item, expandedState, [] );
+				return true;
+			} );
+
+			// New scroll value is the number of expanded items
+			// multiplied by the item height
+			// plus the number of expanded children in the selected block
+			// multiplied by the item height.
+			const newScrollTopValue =
+				listItemHeightFactor * blockListItemHeight +
+				( parentClientIds ? parentClientIds.length : 1 ) *
+					blockListItemHeight;
+
+			const shouldScrollDown =
+				newScrollTopValue >
+				scrollContainer.scrollTop + scrollContainer.clientHeight;
+
+			const shouldScrollUp =
+				newScrollTopValue < scrollContainer.scrollTop;
+
+			if ( ! shouldScrollUp && ! shouldScrollDown ) {
+				return;
+			}
+
+			// @TODO This doesn't yet work when nested blocks are selected.
+			// We're still using the top parent block to calculate/trigger redraw.
+			// If selected block is already visible in the list prevent scroll.
+			scrollContainer?.scrollTo( {
+				top: newScrollTopValue,
```

Comment:
This is not working quite right for some deeply-nested blocks. 
Just making a note for now, will follow it up (and clean it up) later.

![Jan-11-2022 08-59-50](https://user-images.githubusercontent.com/6458278/148845877-b88abf26-110c-45dc-ab40-c7667f4fb87f.gif)